workers/unity/Packages/io.improbable.gdk.core/Systems/EntityReservationSystem.cs

```diff
@@ -0,0 +1,140 @@
+
+            // Request new entity id's? (Maybe move to another system)
+            RequestQueueRefill();
+        }
+
+        private void RequestQueueRefill()
+        {
+            var requestCount = TargetEntityIdCount + queuedReservationCount - inFlightCount - entityIdQueue.Count;
+            if (requestCount > 0)
+            {
+                requestCount = Math.Max(requestCount, MinimumReservationCount);
+                inFlightCount += requestCount;
+
+                var reserveEntityIdsRequest = new WorldCommands.ReserveEntityIds.Request((uint) requestCount);
+                var requestId = commandSystem.SendCommand(reserveEntityIdsRequest);
+                requestIds.Add(requestId);
+            }
+        }
+
+        private void HandleReservationResponses()
+        {
+            var incomingRequests = commandSystem.GetResponses<WorldCommands.ReserveEntityIds.ReceivedResponse>();
+            for (var i = 0; i < incomingRequests.Count; i++)
+            {
+                ref readonly var request = ref incomingRequests[i];
+                if (!requestIds.Contains(request.RequestId))
+                {
+                    continue;
+                }
+
+                if (request.StatusCode == StatusCode.Success)
+                {
+                    //Add range to queue
+                    var range = new EntityRangeCollection.EntityIdRange(request.FirstEntityId.Value, (uint) request.NumberOfEntityIds);
+                    entityIdQueue.Add(range);
+                }
+
+                inFlightCount -= request.RequestPayload.NumberOfEntityIds;
+
+                // Remove ID from the set as it has been handled.
+                requestIds.Remove(request.RequestId);
+            }
+        }
+
+        private void ResolveQueuedRequests()
+        {
+            while (queuedReservations.Count > 0)
+            {
+                var (taskCompletionSource, count) = queuedReservations.Peek();
+                if (taskCompletionSource.Task.IsCanceled)
+                {
+                    // Remove canceled tasks
+                    queuedReservationCount -= count;
+                    queuedReservations.Dequeue();
+                }
+                else if (entityIdQueue.Count < count)
+                {
+                    // Early out if we don't have enough id's yet
+                    break;
+                }
+
+                queuedReservationCount -= count;
+                queuedReservations.Dequeue();
+                taskCompletionSource.SetResult(entityIdQueue.Take(count));
+            }
+        }
+
+        public bool TryGet(out EntityId entityId)
+        {
+            if (entityIdQueue.Count > 0)
+            {
+                entityId = entityIdQueue.Dequeue();
+                return true;
+            }
+
+            entityId = default;
+            return false;
+        }
+
+        public Task<EntityId[]> Take(uint count, CancellationToken cancellationToken = default)
```

Comment:
In terms of API symmetry we have:

- `TryGet` sync
- `Take` async

What about:

- `Get` async
- ` TryTake` sync